### PR TITLE
feat(tactic/interactive): add `lean.parser.itactic`, a tactic block parser

### PR DESCRIPTION
**Pull request by khoek**
Lean handles parsing of most arguments to interactive tactics robustly, using the `parser` monad. However, due to the way that parsing nested tactic blocks and the `begin ... end` syntax is implemented, there is no `parser (tactic unit)` for interactive blocks. Instead arguments to interactive functions of type `tactic.interactive.itactic` are interpreted requesting that an interactive block be parsed.

~~This isn't so bad, since one typically doesn't want to specify as an argument to an interactive tactic a complicated series of multiple nested tactic blocks, as opposed to other data like `name`s for `rewrite`ing. One unfortunate disadvantage however is that we miss-out on being able to write tactics which optionally parse a nested tactic block, which leads to a yuck `{}` syntax appended on the end of tactic applications. Applications include libraries such as [khoek/libiconfig](https://github.com/khoek/libiconfig), which can handle parsing potentially very complicated nested configuration directives and arguments (including passing partially assembled structures) using a custom interaction monad (where the unsightly `{}` will no longer be required).~~

We generalise (in a backwards-compatible way) the `interactive.parse` syntax for parsing things which do not have a `has_reflect` in general, and add the `lean.parser.itactic : parser (tactic unit)` for use in interactive tactics.

### CODE REVIEW

**Comment by digama0** [comment]
Do you understand the reason why `itactic` is not a `parser (tactic unit)`? I would certainly prefer that if possible.

**Comment by khoek** [comment]
I'll look into if it's (easily) possible.

**Comment by khoek** [comment]
@digama0 one real problem is that there is no instance of `reflected (tactic unit)`.

**Comment by khoek** [comment]
The definition is
````
@[reducible] meta def parse {α : Type} [has_reflect α] (p : parser α) : Type := α
````
Do you understand why `[has_reflect α]`, and why it is used internally? I can happily write
````lean
variables (α : Type) [inhabited α]
meta def ppp : lean.parser α := return $ inhabited.default α 
````
but I'll never be able to `parse ppp`?

**Comment by khoek** [comment]
Mario made some comments on Zulip: https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/has_reflect.20in.20interactive.2Eparse

I'll implement that

**Comment by khoek** [comment]
@digama0 There we go, implemented!

**Comment by khoek** [comment]
Rebased on master, also better test

**Comment by khoek** [comment]
The way we implement `parse_interactive_param` now actualy considerably simplifies the logic.

File: library/init/meta/lean/parser.lean [hunk 1]
```diff
@@ -85,5 +107,19 @@ meta constant of_tactic : tactic α → parser α
...
+
+meta instance cast (p : lean.parser (reflected_value α)) : reflectable (val p) :=
+{full := p}
+
+meta instance has_reflect [r : has_reflect α] (p : lean.parser α) : reflectable p :=
+{full := do rp ← p, return (reflected_value.reflect rp)}
+
+meta instance optional {α : Type} [reflected α] {p : parser (reflected_value α)}
+    : reflectable (optional (val p)) :=
+{full := reflected_value.subst some <$> p <|> return (reflected_value.mk none)}
```
Comment (digama0) [inline library/init/meta/lean/parser.lean]: I think you should have an instance of `reflectable (val p)` that uses `p`... That way you can have the `optional` instance say `[reflectable p] : reflectable (optional p)`, which is what I expected to see. Or is there another reason you wrote it that way?

Comment (khoek) [reply library/init/meta/lean/parser.lean]: That is instance is the content of `reflectable.cast`, and sure that sounds better.

**Comment by khoek** [comment]
Everything compiles, but just one more thing---I'm going to rename `[r : reflected val]` to `[reflect : reflected val]`

**Comment by khoek** [comment]
There we go, done

**Comment by khoek** [comment]
Merge me? :D :dancer:

**Pull request merged**